File: 105.ConstructBinaryTreeFromInorderAndPreorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def __init__(self):
        self.indexRef = {}

    def build(self, preorder, ilb, irb, plb, prb):
        if ilb > irb:
            return None

        root = TreeNode(preorder[plb])
        mid = self.indexRef[root.val]
        root.left = self.build(preorder, ilb, mid - 1, plb + 1, plb + 1 + (mid - 1 -ilb))
        root.right = self.build(preorder, mid + 1, irb, prb - (irb - (mid + 1)), prb)
        return root

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        for i in range(len(inorder)):
            self.indexRef[inorder[i]] = i
        
        return self.build(preorder, 0, len(inorder) - 1, 0, len(preorder) - 1)

File: 105.ConstructBinaryTreeFromInorderAndPreorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    Map<Integer, Integer> indexRef = new HashMap<>();

    TreeNode build(int[] preorder, int ilb, int irb, int plb, int prb) {
        if (ilb > irb) {
            return null;
        }

        TreeNode root = new TreeNode(preorder[plb]);
        int mid = indexRef.get(root.val);
        root.left = build(preorder, ilb, mid - 1, plb + 1, plb + 1 + (mid - 1 - ilb));
        root.right = build(preorder, mid + 1, irb, prb - (irb - (mid + 1)), prb);
        return root;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i=0; i<inorder.length; i++) {
            indexRef.put(inorder[i], i);
        }
        return build(preorder, 0, inorder.length - 1, 0, preorder.length - 1);
    }
}
